File: my_ai/alpha_beta.py
import time
import sys
from copy import deepcopy
from move import Move
from position import Position

class AlphaBeta():
    def __init__(self, evaluator, time):
        self.__time_limit = time - 0.015
        self.__evaluator = evaluator
        self.__get_valid_moves = 0
        self.__deepcopies = 0
        self.__make_moves = 0
        self.__recursions = 0

    def __alpha_beta(self, initial_position, depth, alpha, beta, maximizing):
        self.__recursions += 1
        if depth == 0:
            evaluation_move = Move(value=self.__evaluator.evaluate(initial_position))
            #print(initial_position, file=sys.stderr)
            #print('evaluation move: ' + str(evaluation_move) + ', value: ' + str(evaluation_move.value), file=sys.stderr)
            return evaluation_move
        start = time.process_time()
        valid_moves = initial_position.get_valid_moves()
        self.__get_valid_moves += time.process_time() - start
        if len(valid_moves) == 0:
            evaluation_move = Move(value=self.__evaluator.evaluate(initial_position))
            #print(initial_position, file=sys.stderr)
            #print('evaluation move: ' + str(evaluation_move) + ', value: ' + str(evaluation_move.value), file=sys.stderr)
            return evaluation_move
        if maximizing:
            max_value = -float('inf')
            best_move = Move()
            for move in valid_moves:
                start = time.process_time()
                new_position = deepcopy(initial_position)
                #new_position = Position()
                #new_position.set_grid(initial_position.get_grid())
                #new_position.set_player(initial_position.get_player())
                self.__deepcopies += time.process_time() - start
                start = time.process_time()
                new_position.make_move(move)
                self.__make_moves += time.process_time() - start
                result_move = self.__alpha_beta(new_position, depth - 1, alpha, beta, False)
                #if depth == 3:
                #    print('Max, depth: ' + str(depth) + ', move: ' + str(move) + ' ' + str(result_move.value), file=sys.stderr)
                if result_move.value > max_value:
                    max_value = result_move.value
                    move.value = max_value
                    best_move = move
                if max_value >= beta:
                    move.value = max_value
                    return move
                alpha = max(alpha, max_value)
                #print(alpha, file=sys.stderr)
                #if alpha >= beta:
                    #print('break max', file=sys.stderr)
                #    break
            return best_move
        else:
            min_value = float('inf')
            best_move = Move()
            for move in valid_moves:
                start = time.process_time()
                new_position = deepcopy(initial_position)
                #new_position = Position()
                #new_position.set_grid(initial_position.get_grid())
                #new_position.set_player(initial_position.get_player())
                self.__deepcopies += time.process_time() - start
                start = time.process_time()
                new_position.make_move(move)
                self.__make_moves += time.process_time() - start
                result_move = self.__alpha_beta(new_position, depth - 1, alpha, beta, True)
                #print('Min, depth: ' + str(depth) + ', move: ' + str(move) + ' ' + str(result_move.value), file=sys.stderr)
                if result_move.value < min_value:
                    min_value = result_move.value
                    move.value = min_value
                    best_move = move
                if min_value <= alpha:
                    move.value = min_value
                    return move
                beta = min(beta, min_value)
                #print(beta, file=sys.stderr)
                #if alpha >= beta:
                    #print('break min', file=sys.stderr)
                #    break
            return best_move

    def best_move(self, initial_position):
        if len(initial_position.get_valid_moves()) == 0:
            return Move(is_pass_move=True)#'pass'
        start = time.process_time()
        result = self.__alpha_beta(initial_position, 7, -float('inf'), float('inf'), True)
        print('Total (s): ' + str(time.process_time() - start), file=sys.stderr)
        print('Deepcopies (s): ' + str(self.__deepcopies), file=sys.stderr)
        print('Make moves (s): ' + str(self.__make_moves), file=sys.stderr)
        print('Get valid moves (s): ' + str(self.__get_valid_moves), file=sys.stderr)
        print('Recursions: ' + str(self.__recursions), file=sys.stderr)
        return result

File: java/src/AlphaBeta.java
import java.util.*;

/**
 * This class is used to run the alpha beta algorithm.
 * @author Thomas Ranvier
 */
public class AlphaBeta {
    private double time_limit;
    private double timer;
    private Evaluator evaluator;
    private char player;

    /**
     * Class constructor taking as argument the evaluator to use and the time limit.
     * @param player The player that will be considered as max.
     * @param evaluator Evaluator that will be used to evaluate each game state.
     * @param time_limit The time limit that the algorithm must respects.
     */
    public AlphaBeta(char player, Evaluator evaluator, double time_limit) {
        this.time_limit = time_limit - 0.002;
        this.evaluator = evaluator;
        this.timer = 0;
        this.player = player;
    }

    /**
     * Calls the alpha beta function to determine the best move to do.
     * @param initial_position The initial position of the game.
     * @return The best move determined.
     */
    public Move bestMove(Position initial_position) {
        if (initial_position.getValidMoves().size() == 0)
            return new Move(true);
        this.timer = System.nanoTime();
        int depth = 5;
        Move best_move = new Move(true);
        while (depth < 20 && (System.nanoTime() - this.timer) / 1_000_000_000.0 < this.time_limit) {
            try {
                best_move = this.alphaBeta(initial_position, depth, Integer.MIN_VALUE, Integer.MAX_VALUE);
                System.err.println("Reached depth: " + depth);
                depth++;
            } catch (OutOfTimeException exception) {}
        }
        return best_move;
    }

    /**
     * This function implements the alpha beta algorithm.
     * @param initial_position The game state on which the function is called.
     * @param depth The actual depth.
     * @param alpha The actual value of alpha.
     * @param beta The actual value of beta.
     * @return The local best move depending on how deep the call is in the recursion tree.
     */
    private Move alphaBeta(Position initial_position, int depth, int alpha, int beta)
            throws OutOfTimeException {
        if (depth == 0)
            return new Move(this.evaluator.evaluate(initial_position));
        ArrayList<Move> valid_moves = initial_position.getValidMoves();
        if (valid_moves.size() == 0)
            return new Move(this.evaluator.evaluate(initial_position));
        boolean maximizing = initial_position.getPlayer() == this.player;
        int value = maximizing ? Integer.MIN_VALUE : Integer.MAX_VALUE;
        Move best_move = new Move(0, 0);
        for (Move move: valid_moves) {
            if ((System.nanoTime() - this.timer) / 1_000_000_000.0 > this.time_limit)
                throw new OutOfTimeException();
            Position new_position = initial_position.clone();
            new_position.makeMove(move);
            Move result_move = this.alphaBeta(new_position, depth - 1, alpha, beta);
            boolean better_value = maximizing ? result_move.value > value : result_move.value < value;
            if (better_value) {
                value = result_move.value;
                move.value = value;
                best_move = move;
            }
            boolean cut_search = maximizing ? value >= beta : value <= alpha;
            if (cut_search) {
                move.value = value;
                return move;
            }
            if (maximizing)
                alpha = alpha > value ? alpha : value;
            else
                beta = beta < value ? beta : value;
        }
        return best_move;
    }
}
